lib/everypolitician/popolo/collection.rb

```diff
@@ -31,8 +31,14 @@ def find_by(attributes = {})
       end
 
       def where(attributes = {})
-        select do |object|
-          attributes.all? { |k, v| object.send(k) == v }
+        index_attributes(attributes)
+        attributes.keys.map { |k| @indexes[k.to_sym][attributes[k]] }.reduce(:&) || []
```

Comment:
Generally we prefer avoiding breaking the encapsulation as much as possible, even internally, avoiding direct attribute lookups and going through the method calls each time (unless we actually need to micro-optimise). Here I think it would be cleaner to replace the command method with a simple lookup per attribute:

```ruby
      def index_for(attr)
        (@indexes ||= {}).fetch(attr, group_by(&attr))
      end
```

and then the search would become:

```ruby
      def where(attributes = {})
        attributes.map { |k, v| index_for(k.to_sym)[v] }.reduce(:&) || []
      end
```